tket/src/Circuit/Boxes.cpp

```diff
@@ -494,6 +494,298 @@ op_signature_t StabiliserAssertionBox::get_signature() const {
+  ToffoliBox::cycle_transposition_t return_transposition;
+  for (unsigned k = 0; k < cycle_transpositions.size(); k++) {
+    ToffoliBox::cycle_transposition_t cycle = cycle_transpositions[k];
+    unsigned i = 0, j = 1;
+    while (j < cycle.size()) {
+      ToffoliBox::transposition_t transposition_i = cycle[i];
+      ToffoliBox::transposition_t transposition_j = cycle[j];
+
+      std::vector<bool> transposition_j_first = transposition_j.first;
+      std::vector<bool> transposition_i_last = transposition_i.last;
+      std::vector<bool> transposition_i_first = transposition_i.first;
+
+      TKET_ASSERT(transposition_i_last == transposition_j.first);
+      std::vector<bool> i_middle = transposition_i.middle;
+      std::vector<bool> j_middle = transposition_j.middle;
+
+      TKET_ASSERT(i_middle.size() == transposition_i.last.size());
+      TKET_ASSERT(j_middle.size() == transposition_i.last.size());
+      // if a transposition has already been reduced, still need to make sure we
+      // uncompute it
+      if (transposition_i_first != transposition_i_last) {
+        unsigned middle_last_distance =
+            get_hamming_distance(i_middle, transposition_i_last);
+        unsigned middle_first_distance =
+            get_hamming_distance(i_middle, transposition_i_first);
+        // this => the reduced transposition is on a good gray code between the
+        // new "first" and target
+        if (middle_first_distance < middle_last_distance &&
+            middle_first_distance > 1) {
+          transposition_i_last = transposition_i_first;
+          std::vector<bool> starting_point = transposition_i_last;
+          for (unsigned k = 0; k < i_middle.size(); k++) {
+            if (i_middle[k] == j_middle[k] &&
+                get_hamming_distance(starting_point, i_middle) > 1) {
+              starting_point[k] = i_middle[k];
+            }
+          }
+
+          cycle_transpositions[k][i].last = starting_point;
+          cycle_transpositions[k][j].first = starting_point;
+        }
+      } else {  // else in this case just find any good transposition
+        std::vector<bool> starting_point = transposition_i_last;
+        for (unsigned k = 0; k < i_middle.size(); k++) {
+          if (i_middle[k] == j_middle[k] &&
+              get_hamming_distance(starting_point, i_middle) > 1) {
+            starting_point[k] = i_middle[k];
+          }
+        }
+
+        cycle[i].last = starting_point;
+        cycle[j].first = starting_point;
+      }
+
+      ++i;
+      ++j;
+    }
+
+    return_transposition.insert(
+        return_transposition.end(), cycle.begin(), cycle.end());
+  }
+
+  return return_transposition;
+}
+
+void ToffoliBox::generate_circuit() const {
+  // This decomposition is as described on page 191, section 4.5.2 "Single
+  // qubit and CNOT gates are universal" of Nielsen & Chuang
+  std::vector<ToffoliBox::cycle_transposition_t> cycle_transpositions =
```

Comment:
If we do decide to make `reorder=true` the only option, it would be clearer if `merge_cycles` were moved into the body of this method, rather than defining this vector which then gets mutated and discarded.

Reply:
I've removed `reorder`, so no longer a concern